### PR TITLE
Unit tests for EventDispatcher

### PR DESCRIPTION
**Pull request by squall0gd**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by squall0gd** [comment]
Current coverage ~89%

File: pkg/kubelet/cm/event_dispatcher.go [hunk 1]
```diff
@@ -187,6 +187,7 @@ func (ed *eventDispatcher) Start(socketAddress string) {
 	if err != nil {
 		glog.Fatalf("failed to bind to socket address: %v", err)
 	}
+	fmt.Println(lis.Addr().String())
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher.go]: :) please remove

**Review by flyingcougar — CHANGES_REQUESTED** [review]

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+			t.Errorf("Expected number of isolators (%d) is different than real one (%d)",
+				len(ed.isolators),
+				testCase.requestedIsolatorsNumber,
+			)
+		}
+
+		for registeredIsolator := range ed.isolators {
+			if sort.SearchStrings(testCase.requestedIsolators, ed.isolators[registeredIsolator].name) < 0 {
+				t.Errorf("Isolator: %q has not been found in expected isolators %q",
+					registeredIsolator,
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: add check for socket/name correlation corectness

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+			isolatorName: "test2",
+		},
+		{
+			isolators:    []*lifecycle.RegisterRequest{},
+			isolatorName: "test1",
+		},
+	}
+
+	for _, testCase := range testCases {
+		ed := newEventDispatcher()
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: here is inside for loop, in test case above outside, please be consistent

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+		ed.isolators = make(map[string]*registeredIsolator)
+		for _, isolator := range testCase.isolators {
+			go getIsolators(ed.GetEventChannel(), &EventDispatcherEvent{})
+			ed.Register(ctx.Background(), isolator)
+		}
+		var event EventDispatcherEvent
+		go getIsolators(ed.GetEventChannel(), &event)
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: please also check for event

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+		ed.Unregister(context.Background(), &lifecycle.UnregisterRequest{Name: testCase.isolatorName})
+		if ed.isolators[testCase.isolatorName] != nil {
+			t.Error("Unregistration failed: expected item to remove is still available")
+		}
+	}
+}
+
+type dummy struct {
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: s/dummy/`sth else`

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+			serverError: fmt.Errorf("Server Error"),
+		},
+		{
+			replyError:  "",
+			serverError: nil,
+		},
+	}
+
+	for _, testCase := range testCases {
+		server.ResetDummyObject()
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: please rename ResetDummyObject

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+		server.SetServerError(testCase.serverError)
+
+		eventReply, err := ed.PreStartPod(pod, "/")
+		if err != nil {
+			if !strings.Contains(err.Error(), testCase.serverError.Error()) {
+				t.Errorf("Expected error(%q) isn't message of grpc error(%q)",
+					testCase.serverError,
+					err)
+			}
+			if eventReply.Error != "" {
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: please add comment on that case

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+	}
+
+	for _, testCase := range testCases {
+		server.ResetDummyObject()
+		server.SetReplyError(testCase.replyError)
+		server.SetServerError(testCase.serverError)
+
+		eventReply, err := ed.PreStartPod(pod, "/")
+		if err != nil {
+			if !strings.Contains(err.Error(), testCase.serverError.Error()) {
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: please add comment on that case

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+			serverError: fmt.Errorf("Server Error"),
+		},
+	}
+
+	for _, testCase := range testCases {
+		server.ResetDummyObject()
+		server.SetServerError(testCase.serverError)
+
+		err := ed.PostStopPod("/")
+		if err != nil {
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: please add comment on that case

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+				},
+			},
+			resources:     &ResourceConfig{},
+			shouldBe:      "3",
+			requestedKind: lifecycle.IsolationControl_CGROUP_CPUSET_MEMS,
+		},
+	}
+
+	for _, testCase := range testCases {
+		er := &lifecycle.EventReply{
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: please rename to sth else, more descriptive

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+			IsolationControls: testCase.isolators,
+		}
+		ed := newEventDispatcher()
+		output := ed.ResourceConfigFromReplies(er, testCase.resources)
+
+		switch testCase.requestedKind {
+		case lifecycle.IsolationControl_CGROUP_CPUSET_CPUS:
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: to consider: instead doing `switch case` here, adjust the `testCases` to consist of output from isolators and expected `ResourceConfig`. the in the loop you just need to compare those. The advantage is that adding new test cases involves only appending to `testCase` structure.

File: pkg/kubelet/cm/event_dispatcher_test.go [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+		}
+		if server.GetLastEventFromServer() == nil {
+			t.Error("Lifecycle object should be accessable in lifecycle handler, but it wasn't")
+		}
+
+	}
+}
+
+func TestEventDispatcher_ResourceConfigFromReplies(t *testing.T) {
+	testCases := []struct {
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher_test.go]: i.e. s/testCase/testResourceConfig
sth more descriptive

**Review by flyingcougar — CHANGES_REQUESTED** [review]

File: pkg/kubelet/cm/event_dispatcher_test.go (hunk 2 above)
Comment (squall0gd) [reply pkg/kubelet/cm/event_dispatcher_test.go]: done

File: pkg/kubelet/cm/event_dispatcher_test.go (hunk 4 above)
Comment (squall0gd) [reply pkg/kubelet/cm/event_dispatcher_test.go]: done

File: pkg/kubelet/cm/event_dispatcher_test.go (hunk 3 above)
Comment (squall0gd) [reply pkg/kubelet/cm/event_dispatcher_test.go]: done

File: pkg/kubelet/cm/event_dispatcher_test.go (hunk 5 above)
Comment (squall0gd) [reply pkg/kubelet/cm/event_dispatcher_test.go]: done

File: pkg/kubelet/cm/event_dispatcher_test.go (hunk 6 above)
Comment (squall0gd) [reply pkg/kubelet/cm/event_dispatcher_test.go]: done

File: pkg/kubelet/cm/event_dispatcher_test.go (hunk 8 above)
Comment (squall0gd) [reply pkg/kubelet/cm/event_dispatcher_test.go]: done

File: pkg/kubelet/cm/event_dispatcher_test.go (hunk 7 above)
Comment (squall0gd) [reply pkg/kubelet/cm/event_dispatcher_test.go]: done

File: pkg/kubelet/cm/event_dispatcher_test.go (hunk 12 above)
Comment (squall0gd) [reply pkg/kubelet/cm/event_dispatcher_test.go]: Disagree. Let's be consistent with Kubernetes

File: pkg/kubelet/cm/event_dispatcher_test.go (hunk 10 above)
Comment (squall0gd) [reply pkg/kubelet/cm/event_dispatcher_test.go]: done

**Comment by squall0gd** [comment]
Current coverage > 90%

**Comment by flyingcougar** [comment]
lgtm from me, please check if tests with master branch.

**Review by ConnorDoyle — COMMENTED** [review]
Please squash to 1 commit before merging

**Pull request merged**